receiver/awsfirehosereceiver/metrics_receiver.go

```diff
@@ -36,43 +40,62 @@ var _ firehoseConsumer = (*metricsConsumer)(nil)
 func newMetricsReceiver(
 	config *Config,
 	set receiver.Settings,
-	unmarshalers map[string]unmarshaler.MetricsUnmarshaler,
 	nextConsumer consumer.Metrics,
 ) (receiver.Metrics, error) {
-	recordType := config.RecordType
-	if recordType == "" {
-		recordType = defaultMetricsRecordType
-	}
-	configuredUnmarshaler := unmarshalers[recordType]
-	if configuredUnmarshaler == nil {
-		return nil, fmt.Errorf("%w: recordType = %s", errUnrecognizedRecordType, recordType)
-	}
-
-	mc := &metricsConsumer{
-		consumer:    nextConsumer,
-		unmarshaler: configuredUnmarshaler,
+	c := &metricsConsumer{
+		config:   config,
+		settings: set,
+		consumer: nextConsumer,
 	}
-
 	return &firehoseReceiver{
 		settings: set,
 		config:   config,
-		consumer: mc,
+		consumer: c,
 	}, nil
 }
 
+func (c *metricsConsumer) Start(_ context.Context, host component.Host) error {
+	encoding := c.config.Encoding
+	if encoding == "" {
+		encoding = c.config.RecordType
+		if encoding == "" {
+			encoding = defaultMetricsEncoding
+		}
+	}
+	switch encoding {
+	case cwmetricstream.TypeStr:
+		// TODO: make cwmetrics an encoding extension
+		c.unmarshaler = cwmetricstream.NewUnmarshaler(c.settings.Logger)
+	case otlpmetricstream.TypeStr:
+		// TODO: make otlp_v1 an encoding extension
+		c.unmarshaler = otlpmetricstream.NewUnmarshaler(c.settings.Logger)
+	default:
+		unmarshaler, err := loadEncodingExtension[pmetric.Unmarshaler](host, encoding, "metrics")
+		if err != nil {
+			return err
+		}
+		c.unmarshaler = unmarshaler
+	}
+	return nil
+}
+
 // Consume uses the configured unmarshaler to deserialize the records into a
 // single pmetric.Metrics. If there are common attributes available, then it will
 // attach those to each of the pcommon.Resources. It will send the final result
 // to the next consumer.
-func (mc *metricsConsumer) Consume(ctx context.Context, records [][]byte, commonAttributes map[string]string) (int, error) {
-	md, err := mc.unmarshaler.Unmarshal(records)
-	if err != nil {
-		return http.StatusBadRequest, err
+func (c *metricsConsumer) Consume(ctx context.Context, records [][]byte, commonAttributes map[string]string) (int, error) {
+	metrics := pmetric.NewMetrics()
+	for _, record := range records {
+		recordMetrics, err := c.unmarshaler.UnmarshalMetrics(record)
+		if err != nil {
+			return http.StatusBadRequest, err
+		}
+		metrics = expmetrics.Merge(metrics, recordMetrics)
```

Comment:
I think this is more likely to behave correctly than the logs implementation, but please also ensure this is tested.

Reply:
Done.